### ISSUE TITLE
Need help to convert OSNet in order run Edge TPU

### ISSUE THREAD

**Issue by hoangminhtoan** [google-coral/edgetpu#290]
I need your help to convert the trained OSNet model from
+ Pytorch: https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO
+ Onnx: https://drive.google.com/uc?id=1dJ0ST2lgNlf1OrN4_PkRrWIN5FnH2nrE osnet_ain_x1_0_msmt17.onnx

input_shape [batch_size=16, 3, 256, 128]

Currently, I can covert Pytorch -> Onnx -> TFlite (not full int8), however, I am still stuck at converting to TFLite int8.
I tried to run  tf1.15 and tf2.2, tf2.3, tf.24 for the converter but no luck.

**Comment by Namburger** [comment]
@hoangminhtoan hello, we don't support `pytorch -> onnx -> tflite` but if you share your progression maybe I can give some pointers. Also, this would be a more appropriate place to ask this type of questions:
https://github.com/tensorflow/tensorflow/issues

**Comment by parthjdoshi** [comment]
@Namburger I am following a similar path and managed to get a full integer quantized tflite model as per [this](https://www.tensorflow.org/lite/performance/post_training_integer_quant#convert_using_integer-only_quantization) and [this](https://qiita.com/PINTO/items/865250ee23a15339d556#4-2-9-quantization-from-a-model-for-pytorch). However, when I try to compile it, it fails with the following error:
`Edge TPU Compiler version 15.0.340273435
Invalid model: osnet_ain_x1_0_msmt17_market1501_full_integer_quant.tflite
Model not quantized
`

Could you please help me out?

**Comment by parthjdoshi** [comment]
Yeah, I realised my mistake. The TFlite model isn't fully quantized due to the `SQUARED_DIFFERENCE` and `RSQRT` operators. Any tips for this?

**Comment by parthjdoshi** [comment]
Sorry for commenting multiple times.
I tried using tf-nightly instead of tf 2.4 and I was able to get the model fully quantized, even with the `SQUARED_DIFFERENCE` and `RSQRT` operators.
```
def representative_dataset_gen():
	for image in raw_image_paths:
		image = cv2.imread(image)
		image = cv2.resize(image, (128, 256))
		image = image[..., ::-1]
		image = image * (1 / 255)
		image[0, ...] = (image[0, ...] - 0.485) / 0.229
		image[1, ...] = (image[1, ...] - 0.456) / 0.224
		image[2, ...] = (image[2, ...] - 0.406) / 0.225
		image = tf.expand_dims(tf.convert_to_tensor(image, dtype = tf.float32), axis = 0)
		yield [image]


with open('selected_files.txt') as f:
	raw_image_paths = f.read().split('\n')[:-1]

converter = tf.lite.TFLiteConverter.from_saved_model('/home/Trial_1/osnet_ain_x1_0_msmt17_saved_model')
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
converter.allow_custom_ops = True
converter.representative_dataset = representative_dataset_gen
tflite_quant_model = converter.convert()
with open('osnet_ain_x1_0_msmt17_market1501_full_integer_quant.tflite', 'wb') as w:
	w.write(tflite_quant_model)
```
The file generated is shown below.
[osnet_ain_x1_0_msmt17_market1501_full_integer_quant.tflite.zip](https://github.com/google-coral/edgetpu/files/5786706/osnet_ain_x1_0_msmt17_market1501_full_integer_quant.tflite.zip)

However, I faced a new issue while compiling for the EdgeTPU. 
```
Edge TPU Compiler version 15.0.340273435
ERROR: Didn't find op for builtin opcode 'SQUARED_DIFFERENCE' version '2'

ERROR: Registration failed.

Invalid model: osnet_ain_x1_0_msmt17_market1501_full_integer_quant.tflite
Model could not be parsed
```

**Comment by parthjdoshi** [comment]
From what I have been reading online, I guess these operations aren't supported by the Coral USB Accelerator. However, wouldn't it lead to a custom op or something which would then run on the CPU and not on the Coral?

**Comment by hoangminhtoan** [comment]
> From what I have been reading online, I guess these operations aren't supported by the Coral USB Accelerator. However, wouldn't it lead to a custom op or something which would then run on the CPU and not on the Coral?

I think so, I trained the OSNet model with Tensorflow v.1.15.4, the trained model could easily be converted to tflite int8 with Tensorflow >= 2.3. However, the full quantized model was said not full quantized when compiling with edgtpu_compiler.

**Comment by parthjdoshi** [comment]
@hoangminhtoan What sort of output were you getting when you ran the edgetpu compiler? Could you please share that? Was it the same as the one I have posted above?

**Comment by parthjdoshi** [comment]
Also, did you use `converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]` when converting?

**Comment by hoangminhtoan** [comment]
> Also, did you use `converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]` when converting?

@arsenal-2004  Here is my converter code 

onset was trained follow this repo:
[osnet_keras](https://github.com/jkjung-avt/keras_imagenet/blob/master/models/osnet.py)

[tensorflow_convert_example](https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_classification_ptq_tf2.ipynb#scrollTo=w9ydAmHGHUZl)

```
import tensorflow.compat.v1 as tf 

model = tf.keras.models.load_model('/content/osnet-ckpt-040.h5')

# A generator that provides a representative dataset
def representative_data_gen():
  dataset_list = tf.data.Dataset.list_files(flowers_dir + '/*/*')
  for i in range(1):
    image = next(iter(dataset_list))
    image = tf.io.read_file(image)
    image = tf.io.decode_jpeg(image, channels=3)
    image = tf.image.resize(image, [256, 128])
    image = tf.cast(image / 255., tf.float32)
    image = tf.expand_dims(image, 0)
    yield [image]



converter = tf.lite.TFLiteConverter.from_keras_model_file('/content/osnet-ckpt-040.h5')

# This enables quantization
converter.optimizations = [tf.lite.Optimize.DEFAULT]
# This sets the representative dataset for quantization
converter.representative_dataset = representative_data_gen
# This ensures that if any ops can't be quantized, the converter throws an error
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
# For full integer quantization, though supported types defaults to int8 only, we explicitly declare it for clarity.
converter.target_spec.supported_types = [tf.int8]
# These set the input and output tensors to uint8 (added in r2.3)
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8

tflite_model = converter.convert()

with open('/content/osnet_quant.tflite', 'wb') as f:
  f.write(tflite_model)
```


Model size:
float32: (30MB)
tflite (10MB) 
tflite-int8 : osnet_quant.tflite (3MB)


then I used edgetpu_compiler

```
Edge TPU Compiler version 15.0.340273435 
Invalid model: osnet_quant.tflite Model not quantized
```

**Comment by Namburger** [comment]
Could you guys attach the model here?
@Naveen-Dodda Could you run it to our compiler to see what error message is present?

**Comment by hoangminhtoan** [comment]
Hi @Namburger ,

Thanks for your quick reply,
Here are models

[keras_model](https://drive.google.com/file/d/1i5RIvtZNJL0qKQoRkDxY_pD32XuDSUmW/view?usp=sharing)
[osnet.tflite](https://drive.google.com/file/d/12X1RTUVXxTuYwjfAh8MzX_SCw0DmhdF1/view?usp=sharing)
[osnet_quant.tflite](https://drive.google.com/file/d/1m3Vq3YVHB14_6WDEaWoC96LHmkTAJk7M/view?usp=sharing)

**Comment by parthjdoshi** [comment]
@Namburger , thank you for your help
This is the model generated using TF-nightly
> The file generated is shown below.
> [osnet_ain_x1_0_msmt17_market1501_full_integer_quant.tflite.zip](https://github.com/google-coral/edgetpu/files/5786706/osnet_ain_x1_0_msmt17_market1501_full_integer_quant.tflite.zip)

The original Saved_Model is [here](https://drive.google.com/file/d/17yO7nRgjczY2RvjDfqomMAioOD7d5GdF/view?usp=sharing)

**Comment by parthjdoshi** [comment]
@hoangminhtoan You could try `converter.allow_custom_ops = True` as well

**Comment by parthjdoshi** [comment]
@Namburger @Naveen-Dodda Any updates on this? Could you please help me with this? Or, do you know of any person reidentification models trained for the EdgeTPU?

**Comment by hoangminhtoan** [comment]
> @hoangminhtoan You could try `converter.allow_custom_ops = True` as well

@arsenal-2004, thanks for your suggestion, I tried but it didn't work when compiling, still "Model not quantized"

**Comment by hoangminhtoan** [comment]
@arsenal-2004 @Namburger @Naveen-Dodda

Hi, I now successfully convert to the full quantization the TPU model. I follow this [tutorial](https://colab.research.google.com/github/google-coral/tutorials/blob/master/fix_conversion_issues_ptq_tf2.ipynb#scrollTo=6MPyPU7RH7lO)

The model is not quantized because it received dynamic batch size, so I fix the number of batch size is 1 as the above tutorial, and it works!

**Comment by parthjdoshi** [comment]
@hoangminhtoan That's nice to hear. However, I already have a fixed batch size of 16.
I think the issue with my model is the Instance Normalization layer, so I am trying one without that layer.
The error that I get now is 
```
loc("model/depthwise_conv2d_3/depthwise"): error: Invalid argument: Quantized tensors must have non-zero scales
error: could not translate function : Quantized tensors must have non-zero scales

Internal compiler error. Aborting! 
```

**Comment by parthjdoshi** [comment]
@Namburger @Naveen-Dodda Can you guys try out with this new model as well? Since it uses batch normalisation instead of instance normalisation and throws a different error while compiling, as shown above.
[osnet_x0_25_msmt17.zip](https://github.com/google-coral/edgetpu/files/5794658/osnet_x0_25_msmt17.zip)

**Comment by parthjdoshi** [comment]
@Namburger @Naveen-Dodda I tried a model with batch 1 as well. Faced the same issue.

**Comment by hoangminhtoan** [comment]
@arsenal-2004 

I got the same problem as yours when changing batch_size from 1 to 16. How about you convert the model from osnet_imagenet from [model_zoo](https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO). I haven't checked with pretrained pytorch model yet, that's why I retrained osnet with imagenet in keras tf1.15

**Comment by parthjdoshi** [comment]
@hoangminhtoan The Model_Zoo models have a batch size of 16. I did manage to get an OSNet model working, which did not contain the Instance Normalization layer. I followed the suggestions from this TensorFlow [issue](https://github.com/tensorflow/tensorflow/issues/46334). Thanks for your help :)

**Comment by hoangminhtoan** [comment]
> @hoangminhtoan The Model_Zoo models have a batch size of 16. I did manage to get an OSNet model working, which did not contain the Instance Normalization layer. I followed the suggestions from this TensorFlow [issue](https://github.com/tensorflow/tensorflow/issues/46334). Thanks for your help :)

@arsenal-2004. So did you successfully compile OSNet model and run it on the board? If so, could you provide your conversion code with specific configurations? Thanks!

**Comment by parthjdoshi** [comment]
@hoangminhtoan Yes, I compiled the OSNet model and ran it on the Coral USB Accelerator. I am using OSNet as a feature extractor for person-reid.
1. I used [the OSNet_x0_25](https://drive.google.com/uc?id=1MLC2lKnQvAQgBKZP1EXB6UdmqujY9qVd) model from [this](https://github.com/GeekAlexis/FastMOT/blob/master/scripts/download_models.sh) repo. I previously tried with the InstanceNormalization model from the same repo, but that failed, as can be seen from the earlier comments in this issue.
2. These models have an explicit batch size of 16, and I required a batch size of 1 for my application, so I used [this issue](https://github.com/onnx/onnx/issues/2182) to change the batch size of the ONNX models.
3. These models are in ONNX format, so I converted these to TF Saved_Models using [this great blog](https://qiita.com/PINTO/items/ed06e03eb5c007c2e102#6-8-onnx---openvino-ir-conversion)
4. I then converted the Saved_Model to a fully integer quantized model using the following code.
```
import cv2
import tensorflow as tf
import numpy as np

def representative_dataset_gen():
	for image in raw_image_paths:
		image = cv2.imread(image)
		image = cv2.resize(image, (128, 256))
		image = image[..., ::-1]
		image = image * (1 / 255)
		image[0, ...] = (image[0, ...] - 0.485) / 0.229
		image[1, ...] = (image[1, ...] - 0.456) / 0.224
		image[2, ...] = (image[2, ...] - 0.406) / 0.225
		image = tf.expand_dims(tf.convert_to_tensor(image, dtype = tf.float32), axis = 0)
		yield [image]

with open('selected_files.txt') as f:
	raw_image_paths = f.read().split('\n')[:-1]

# Full Integer Quantization - Input/Output=float32
converter = tf.lite.TFLiteConverter.from_saved_model('/home/parth/Trial_4/saved_model')
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
converter._experimental_new_quantizer = True
converter.allow_custom_ops = True
converter.representative_dataset = representative_dataset_gen
tflite_quant_model = converter.convert()
with open('v_2_osnet_x0_25_msmt17_batch_1_market1501_full_integer_quant.tflite', 'wb') as w:
	w.write(tflite_quant_model)
print("Full Integer Quantization complete! - osnet_x0_25_msmt17_batch_1_market1501_full_integer_quant.tflite")
```
**The main step was using the new experimental quantizer, so make sure that you are using tf-nightly.**
5. I was then able to use the Edge TPU compiler successfully and all operations were mapped to the TPU.

**Comment by hoangminhtoan** [comment]
@arsenal-2004 

Thanks for your help. I will close this topic as the model is fully converted!

**Comment by FerreJanssen** [comment]
@arsenal-2004 Is it possible to share your conversion? I am trying to replicate your process but am running into issues with the conversion from onnx to tensorflow.

**Comment by parthjdoshi** [comment]
@FerreJanssen Sure. I was using Google Colab for this.
1) I installed OpenVINO using the code from [this notebook](https://colab.research.google.com/drive/1xla23daYYbTIfbdHF0nyHzHyoAvVtyaG#scrollTo=PO2uvGWy7oEi).
2) I ran the model optimizer on the Onnx file which I had modified to have a batch size of 1.
```
!python3 /opt/intel/openvino_2021.2.185/deployment_tools/model_optimizer/mo.py \
  --input_model '/content/drive/MyDrive/osnet_x0_25_msmt17_batch_1.onnx' \
  --input_shape [1,3,256,128] \
  --output_dir '/content/drive/MyDrive/Trial_4' \
  --data_type FP32
```
3) After installing the latest version of openvino2tensorflow, I ran the following command.
```
!source /opt/intel/openvino_2021.2.185/bin/setupvars.sh && openvino2tensorflow \
  --model_path '/content/drive/MyDrive/Clutterbot/Trial_4/osnet_x0_25_msmt17_batch_1.xml' \
  --model_output_path '/content/drive/MyDrive/Clutterbot/Trial_4/osnet_x0_25_msmt17_batch_1' \
  --output_saved_model True
```
You can see the converted model [here](https://drive.google.com/drive/folders/1-u4as6PDBhLSpsRjcotf6J269PJvsJN8?usp=sharing).

**Comment by angelcarro** [comment]
@parthjdoshi Have you verified the correct output of the model comparing it with the original?
I have performed the conversion of other OSNet models but the outputs are very different from the original. It seems that the openvino2tensorflow utility has a bug: https://github.com/PINTO0309/openvino2tensorflow/issues/24

**Comment by parthjdoshi** [comment]
No, I did not compare it to the original. However, I have been training an
OSNet model on custom data, which I'll be converting soo. So, I'll let you
know if I have any issues with that.

**Comment by angelcarro** [comment]
@parthjdoshi Have you finally compared your trained OSNet model with the Edge TPU converted version?

**Comment by parthjdoshi** [comment]
Yes, I did. I didn't have drastically different results but I did suffer a
10% drop in top-1 accuracy

**Comment by angelcarro** [comment]
Do you experience that accuracy drop if you use the floating point version of the converted model?

**Comment by parthjdoshi** [comment]
I haven't tested that out. I'll do that and get back to you.

**Comment by abdulw976** [comment]
> I need your help to convert the trained OSNet model from
> 
> * Pytorch: https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO
> * Onnx: https://drive.google.com/uc?id=1dJ0ST2lgNlf1OrN4_PkRrWIN5FnH2nrE osnet_ain_x1_0_msmt17.onnx
> 
> input_shape [batch_size=16, 3, 256, 128]
> 
> Currently, I can covert Pytorch -> Onnx -> TFlite (not full int8), however, I am still stuck at converting to TFLite int8.
> I tried to run tf1.15 and tf2.2, tf2.3, tf.24 for the converter but no luck.

Hi,
This pretrained model uses fixed batch size of 16, but the number of persons detected in the frame may vary.
Suppose I have 13 persons in the frame, then I will get the input shape of  [batch_size=13, 3, 256, 128]. Currently I am getting error due to this. How to handle this.??
